rootfs/etc/nginx/lua/test/balancer/sticky_test.lua

```diff
@@ -0,0 +1,151 @@
+        context("when backend specifies hash function", function()
+            it("should return an instance with the corresponding hash implementation", function()
+                local sticky_balancer_instance = sticky:new(test_backend)
+                local test_backend_hash_fn = test_backend.sessionAffinityConfig.cookieSessionAffinity.hash
+                local test_backend_hash_implementation = util[test_backend_hash_fn .. "_digest"]
+                assert.equal(sticky_balancer_instance.digest_func, test_backend_hash_implementation)
+            end)
+        end)
+
+        context("when backend does not specify hash function", function()
+            it("should return an instance with the default implementation (md5)", function()
+                local temp_backend = util.deepcopy(test_backend)
+                temp_backend.sessionAffinityConfig.cookieSessionAffinity.hash = nil
+                local sticky_balancer_instance = sticky:new(temp_backend)
+                local default_hash_fn = "md5"
+                local default_hash_implementation = util[default_hash_fn .. "_digest"]
+                assert.equal(sticky_balancer_instance.digest_func, default_hash_implementation)
+            end)
+        end)
+    end)
+
+    describe("balance()", function()
+        setup(function()
+            util.sha1_digest = function(msg) return msg, false end -- Don't hash for simplification
+        end)
+
+        teardown(function()
+            package.loaded["util"] = nil
+            util = require("util")
+        end)
+        
+        before_each(function()
+            package.loaded["resty.chash"] = nil
+            resty_chash = require("resty.chash")
+
+            package.loaded["balancer.sticky"] = nil
+            sticky = require("balancer.sticky")
+
+            resty_chash.new = function(self, nodes)
+                return {
+                    find = function(self, key)
+                        -- cookie's value is the unhashed endpoint since we stubbed the hash function
+                        assert.equal(key, test_backend_endpoint)
+                        return test_backend_endpoint
+                    end,
+                    next = function(self, index)
+                        return test_backend_endpoint
+                    end,
+                    npoints = 1
+                }
+            end
+        end)
+
+        context("when client doesn't have a cookie set", function()
+            it("should pick an endpoint for the client", function()
+                local cookie = require("resty.cookie")
+                cookie.new = function(self) 
+                    return { 
+                        get = function(self, n) return false end,
+                        set = function(self, n) return true end
+                    }
+                end
+                local sticky_balancer_instance = sticky:new(test_backend)
+                local s = spy.on(sticky_balancer_instance.instance, "next")
```

Comment:
Can we avoid this? This test should not know the implementation details of `balance` function. Why don't you assert on the return value of `:balance`?

Reply:
Good point, it currently feels a bit hacky. I like the approach you suggested 👍